src/main/java/com/julienviet/pgclient/impl/pubsub/PgSubscriberImpl.java

```diff
@@ -0,0 +1,320 @@
+      handler.handle(null);
+    }
+  }
+
+  @Override
+  public synchronized boolean closed() {
+    return closed;
+  }
+
+  @Override
+  public synchronized PgConnection actualConnection() {
+    return conn;
+  }
+
+  @Override
+  public synchronized PgSubscriber connect(Handler<AsyncResult<Void>> handler) {
+    if (closed) {
+      closed = false;
+      tryConnect(0, handler);
+    }
+    return this;
+  }
+
+  private void tryConnect(long delayMillis, Handler<AsyncResult<Void>> handler) {
+    if (!connecting) {
+      connecting = true;
+      if (delayMillis > 0) {
+        vertx.setTimer(delayMillis, v -> doConnect(handler));
+      } else {
+        doConnect(handler);
+      }
+    }
+  }
+
+  private void doConnect(Handler<AsyncResult<Void>> completionHandler) {
+    PgConnection.connect(vertx, options, ar -> handleConnectResult(completionHandler, ar));
+  }
+
+  private synchronized void handleConnectResult(Handler<AsyncResult<Void>> completionHandler, AsyncResult<PgConnection> ar1) {
+    connecting = false;
+    if (ar1.succeeded()) {
+      conn = ar1.result();
+      conn.notificationHandler(PgSubscriberImpl.this::handleNotification);
+      conn.closeHandler(PgSubscriberImpl.this::handleClose);
+      if (channels.size() > 0) {
+        List<Handler<Void>> handlers = channels.values()
+          .stream()
+          .flatMap(channel -> channel.subs.stream())
+          .map(sub -> sub.subscribeHandler)
+          .filter(Objects::nonNull)
+          .collect(Collectors.toList());
+        String sql = channels.values()
+          .stream()
+          .map(channel -> {
+            channel.subscribed = true;
+            return channel.name;
+          })
+          .collect(Collectors.joining(";LISTEN ", "LISTEN ", ""));
+        conn.query(sql, ar2 -> {
+          if (ar2.failed()) {
+            conn.close();
+          } else {
+            handlers.forEach(vertx::runOnContext);
+          }
+          completionHandler.handle(ar2.mapEmpty());
+        });
+        return;
+      }
+    }
+    completionHandler.handle(ar1.mapEmpty());
+  }
+
+  private class ChannelList {
+
+    final String name;
+    final ArrayList<ChannelImpl> subs = new ArrayList<>();
+    boolean subscribed;
+
+    ChannelList(String name) {
+      this.name = name;
+    }
+
+    void add(ChannelImpl sub) {
+      subs.add(sub);
+      if (!subscribed) {
+        if (conn != null) {
+          subscribed = true;
+          String sql = "LISTEN " + name;
+          conn.query(sql, ar -> {
+            if (ar.succeeded()) {
+              Handler<Void> handler = sub.subscribeHandler;
+              if (handler != null) {
+                handler.handle(null);
+              }
+            }
+          });
+        }
+      }
+    }
+
+    void remove(ChannelImpl sub) {
+      subs.remove(sub);
+      if (subs.isEmpty()) {
+        channels.remove(name, this);
+        if (conn != null) {
+          conn.query("UNLISTEN " + name, ar -> {});
```

Comment:
Are we sure we want to ignore the result? I guess that errors can be important

Reply:
I think logging is ok; the thing is that I don't see a valid reason why it would fail (at least for now)